### DISCUSSION TITLE
Feedback on the new 4.3 grouped notifications API

### DISCUSSION THREAD

**Discussion by renchap** [mastodon/mastodon#31618] [General]
In Mastodon 4.3, we are adding support for server-side notification grouping.

This is intended to make notifications more usable for people with a lot of interactions. Currently `boost` and `favorite` notifications are grouped, and we want to add an option to also group `follow`.

This required creating new API endpoints for this feature to work. A (not yet 100% finished) documentation is available here: https://docs.joinmastodon.org/methods/notifications_alpha/

This API is available in Mastodon 4.3.0 beta1 as experimental, but we intend to mark it as stable before the final release, after receiving feedback on it and testing its performance more widely on mastodon.social. You can see how they work here: https://oisaur.com/@renchap/112812306472121018

<img width="597" alt="image" src="https://github.com/user-attachments/assets/dc668081-79ee-479b-8774-e59681140d09">

The big changes are:
- notifications are grouped by the server, when they are received. The algorithm for grouping should not be of concern for clients. At the moment, they are grouped such as notifications for the same event received within a 12-hour timespan are part of the same group
- notifications now have an opaque `group_key` attribute. It will be unique for every non-grouped notification, and is sent through the streaming server
- the new endpoint will return notification groups as the main entity. Each group will contain data about the notification, plus a list of the latest accounts that are part of this group (called sampled accounts)
- in a change from the existing endpoints, we do not embed `account` entities into the notification groups, but only their IDs. Account data for all accounts in the response payload are available in an additional root-level key. This is done to avoid sending data about the same account many times in the same response (our real-life testing showed that it deduplicates 30 to 60% of the accounts data, resulting is smaller payload and faster server response)
- you can pass a new `expand_accounts` parameter, to tell the server that you do not want full account information for every sample account, but only for the first one of them, and a much smaller (and faster to generate for the server) "partial account" for the other ones, with just enough data to display them as avatars
- we are considering having another `expand_accounts=none` option here, to tell the server that you do not want any data about those sample accounts, but we would only add this if some client developer think it would be useful to them. You could want to use this because you have some local cache of accounts and do not want to fetch those accounts again, maybe also in combination with [our new endpoint](https://docs.joinmastodon.org/methods/accounts/#index) to get multiple accounts data at once.
- we are considering letting the client specify the types of notification that should be grouped in the response (see https://github.com/mastodon/mastodon/pull/31594). This would allow us to make some types groupable in the future without breaking clients, and allow clients to offer an opt-out feature for some types to their users.

We are looking for client developer feedback on this API to ensure that we make it as usable as possible, and did not forget anything of importance. This API is the basis for our new notification UI in 4.3 beta so we know it works for our usage. If you find it good and can not wait for it to be released as stable, you can also tell us ;)

We want to hear *specifically from developers of Mastodon client applications*, about the API - we are not expecting to modify the way that grouping looks or works directly. We have been through many experiments and feedback already, and there are good reasons for the choices made so far - we don't intend to make major or drastic changes to the API, but please let us know if you think we can improve anything.

**Comment by connyduck** [comment]
The documentation already looks very nice, sadly I won't have time to implement it anytime soon.

What I don't fully understand yet is how pagination works when a notification group goes over the bounds of a page and what exactly the `page_min_id` and `page_max_id` do. Maybe you could add a more detailed example to the documentation?

Also I'm missing the `filtered` parameter that the old notification endpoint has. https://github.com/mastodon/mastodon/pull/31011

> we are considering letting the client specify the types of notification that should be grouped in the response (see https://github.com/mastodon/mastodon/pull/31594). This would allow us to make some types groupable in the future without breaking clients, and allow clients to offer an opt-out feature for some types to their users.

Yes please, that sounds super useful.

**Reply by ClearlyClaire** [reply]
> Also I'm missing the `filtered` parameter that the old notification endpoint has. #31011

Ah, interesting, but then quite tricky! If you include filtered notifications… then some groups can have notifications that are filtered together with notifications that aren't…

**Reply by connyduck** [reply]
How about a group that contains all filtered notifications?

**Reply by ClearlyClaire** [reply]
I suppose we could add something for that, but considering there would be groups containing both filtered and unfiltered notifications… that does seem a bit weird. Or we could have that for notification groups that are in fact ungrouped by nature (e.g. mentions).

**Comment by DagAgren** [comment]
Is there a way to request the exact list of accounts in a group? Most of the APIs seem to return "some" of them, which is useful for showing in a top-level list, but I do want to be able to tap on the list to see exactly who is on the list.

**Reply by ClearlyClaire** [reply]
There is no way to do that yet! But we will consider it!

**Reply by ClearlyClaire** [reply]
We added `/api/v2_alpha/notifications/:group_key/accounts` for this purpose. It will be in the next beta.

**Comment by evant** [comment]
I currently have client-side grouping implemented and this api does seem to have all the information required to display them the same way. 👍 One difference is that I'm grouping boosts and favorites into the same notification. I don't know if that's ever something you'd be interested in doing server-side but I don't know if the single 'type' attribute is enough to support that?

**Reply by ClearlyClaire** [reply]
We are not planning on grouping notifications of different types at the moment.

**Reply by evant** [reply]
fair enough, I was thinking of this on more an api evolution point of view than a feature request. But it's not a big deal in the scheme of things.

**Comment by evant** [comment]
Have you thought about what screenreaders should read out for the avatars? My first thought would be to use the account's display_name which isn't currently returned for PartialAccountWithAvatar, but I don't know if there's an alternative that would work better here.

**Reply by ClearlyClaire** [reply]
One reason we decided to not expose `display_name` is that we thought exposing `display_name` would make the most sense with exposing custom emojis as well (since they can appear in display names), which is a whole other can of worms. But I suppose we can consider exposing `display_name` without custom emojis if that makes sense for screenreaders.

**Comment by Tunous** [comment]
I have a client application that implements groping logic locally. In my applications notifications that happened on different days are displayed on separate pages. Using this api how could I determine for which day a specific group should be displayed?

I can see `latest_page_notification_at` but it is marked as optional and it's not clear for me when it would not be sent. Does "This is only returned when paginating through notification groups." from documentation mean that if I request first page without any pagination parameters the date won't be returned?

**Reply by ClearlyClaire** [reply]
This information is available when requesting `/api/v2_alpha/notifications` but not `/api/v2_alpha/notifications/:group_key`, we'll try to make that clearer.

**Comment by shadowfacts** [comment]
My app implements client-side grouping currently, and there are a couple things missing from the current alpha API that I need to have feature parity:

- An option to include display names and custom emoji in the partial account objects. I [currently](https://vaccor.space/tusker/img/tusker-notif-grouping-dark.png) show the display names of accounts that originated the notifications, including the custom emojis. This information should be included in the grouped notifications response, otherwise I'd have to separately fetch the accounts for each group, turning loading into n+1 HTTP requests (I think this would also be the case with the Mastodon web frontend's design for groups—it's not clear to me if/how that's avoiding this problem).

- A separate endpoint to get _all_ of the accounts associated with a group. My current implementation lets you select a group to [view all of the accounts](https://github.com/user-attachments/assets/0821870d-f628-4c43-bd6e-506682a91138) that performed the particular action, and the alpha API doesn't seem to have a way of recovering that information. If the concern with including that in the main groups endpoint is performance, it would be entirely reasonable to have a separate API endpoint (possible paginated?) for it.

It's also not entirely clear to me how pagination works in the groups API. In actuality, are you still paginating through individual notifications and the API is just returning groups, or are you actually paginating through groups? If the former, is it ever the case that advancing a page in the groups API could return a group that was present in the previous page? And, even if that scenario isn't possible, is "you're paginating through individual notifs under the hood" the reason why there's no `oldest_notification_id` analog to `most_recent_notification_id` (and similarly for `latest_page_notification_at`)? I can imagine scenarios where it'd be useful to have that info (e.g., knowing what time span is covered by a group).

**Reply by ClearlyClaire** [reply]
> An option to include display names and custom emoji in the partial account objects. I [currently](https://vaccor.space/tusker/img/tusker-notif-grouping-dark.png) show the display names of accounts that originated the notifications, including the custom emojis. This information should be included in the grouped notifications response, otherwise I'd have to separately fetch the accounts for each group, turning loading into n+1 HTTP requests

This is what happens if you set `expand_accounts=full` or omit it. 

> (I think this would also be the case with the Mastodon web frontend's design for groups—it's not clear to me if/how that's avoiding this problem).

The current interface does request the full accounts because of how the information is stored, but it would not need to after a rewrite, UX wise: only the first account is shown in full, the others are just a list of avatars (with a hover card, which data can be fetched on request—but a user hovering over all avatars is unlikely in practice)

> A separate endpoint to get _all_ of the accounts associated with a group. My current implementation lets you select a group to [view all of the accounts](https://github.com/user-attachments/assets/0821870d-f628-4c43-bd6e-506682a91138) that performed the particular action, and the alpha API doesn't seem to have a way of recovering that information. If the concern with including that in the main groups endpoint is performance, it would be entirely reasonable to have a separate API endpoint (possible paginated?) for it.

This is indeed not available at the moment. We are considering it.

> It's also not entirely clear to me how pagination works in the groups API. In actuality, are you still paginating through individual notifications and the API is just returning groups, or are you actually paginating through groups? If the former, is it ever the case that advancing a page in the groups API could return a group that was present in the previous page?

We intend to publish a more comprehensive piece of documentation that covers all of this, but basically: we are actually paginating through individual notifications, and the API is just returning groups. There can actually be overlap between pages, although this is somewhat limited in practice.

> And, even if that scenario isn't possible, is "you're paginating through individual notifs under the hood" the reason why there's no `oldest_notification_id` analog to `most_recent_notification_id` (and similarly for `latest_page_notification_at`)? I can imagine scenarios where it'd be useful to have that info (e.g., knowing what time span is covered by a group).

No, it is just that we found no use for this. I am not sure how useful knowing the precise time span covered by a group is.

**Reply by shadowfacts** [reply]
Thanks for the info, this is very helpful!

> This is what happens if you set expand_accounts=full or omit it.

Ah, gotcha. I'm not sure where exactly the tradeoff lies, but it might still be helpful to have a partial variant that includes the display name. (Though I do appreciate that if you go too far down this route, you're basically reinventing GraphQL.)

> The current interface does request the full accounts because of how the information is stored, but it would not need to after a rewrite

Wouldn't showing the display name for only the first account still require either `expand_accounts=full` or doing n+1 queries? Or am I misunderstanding something here.

> This is indeed not available at the moment. We are considering it.

I'm glad to hear it!

> No, it is just that we found no use for this. I am not sure how useful knowing the precise time span covered by a group is.

I'm currently considering the possibility of having a hybrid grouping mode, where some is done on the server and other grouping is performed on the client (e.g., "Person A favorited 5 posts"), for which it would be helpful to know the time span.

**Reply by ClearlyClaire** [reply]
> A separate endpoint to get _all_ of the accounts associated with a group.

We added `/api/v2_alpha/notifications/:group_key/accounts` for this purpose. It will be in the next beta.

**Comment by shaedrich** [comment]
Thank you for implementing this feature! In most cases, this looks amazing! 👍🏻 

However, there's one particular case: Every now and then, posts only really start at the second line. The first line then only contains auto-filled at-mentions and such, so no helpful information.\
Therefore, I wonder if the number of lines shown could be increased by default or made configurable. Another option would be to collapse the line breaks in that preview—what do you think?
![image_2024-08-29_09-40-24](https://github.com/user-attachments/assets/0c41493a-7b18-4f52-acd7-f1ce7d0cab6b)

**Reply by ClearlyClaire** [reply]
Hi! This discussion is about the API, not the user interface! To track this, I would suggest creating a separate issue or discussion.
Thanks for the feedback, though!

EDIT: https://github.com/mastodon/mastodon/issues/31668 seems to correspond to your issue

**Reply by shaedrich** [reply]
Hi! Thanks for pointing this out!

**Comment by mooseyboots** [comment]
hi, i have implemented grouped notifications and it is mostly working well. but i have an issue with handling non-grouped notifs when grouping is enabled.

the concrete case is that i can't fetch or dismiss single ungrouped notifications. no matter what i pass as the `:group-key` in the path params, nothing happens.

i have tried:

- "ungrouped-{notification-id}"
- "{notification-id}"
- "{group-key}"
- "ungrouped-{group-key}"

the docs say the first should work, i think, or maybe the last, but they're confusing (to me).

i'm following this part of the docs: https://docs.joinmastodon.org/methods/grouped_notifications/#query-parameters:

> grouped_types[]
>     Array of String. Restrict which notification types can be grouped. Use this if there are notification types for which your client does not support grouping. If omitted, the server will group notifications of all types it supports (currently, favourite, follow and reblog). If you do not want any notification grouping, use [GET /api/v1/notifications](https://docs.joinmastodon.org/methods/notifications/#get) instead. Notifications that would be grouped if not for this parameter will instead be returned as individual single-notification groups with a unique group_key that can be assumed to be of the form ungrouped-{notification_id}. Please note that neither the streaming API nor the individual notification APIs are aware of this parameter and will always include a “proper” group_key that can be different from what is returned here, meaning that you may have to ignore group_key for such notifications that you do not want grouped and use ungrouped-{notification_id} instead for consistency.

can anyone enlighten me on the right way to handle this?

**Reply by mooseyboots** [reply]
for trying to fetch i get an error: "record not found", for trying to dismiss i get 200, but nothing happens.

**Reply by ClearlyClaire** [reply]
You are not doing anything wrong, I think it is currently not possible to target an ungrouped notification using those two endpoints.

**Reply by mooseyboots** [reply]
hi C, thanks for the clarification. so for any notif types that are ungrouped we just fall back to v1 i guess

**Reply by mooseyboots** [reply]
i wonder then if that part of the docs needs amending? it seems to me to suggest that using `ungrouped-{notification-id}` would work.

**Reply by ClearlyClaire** [reply]
Arguably, this is a bug. https://github.com/mastodon/mastodon/pull/33990

**Reply by mooseyboots** [reply]
ok, glad i asked! i thought it would make sense if we could still wrangle single notifs using v2. and it looks like some provisions were being made to do that. i'll follow the PR.

**Reply by mooseyboots** [reply]
just to confirm, once the next release is cut, i figure that non-grouped notif types will need to use `ungrouped-{notification-id}` for such endpoints? (currently the server seems to return `ungrouped-{group-key}`, ie a 7-digit number, not a standard long notif id.)

**Reply by ClearlyClaire** [reply]
Yes, it should return `ungrouped-{notification.id}`. This is what the server returns. Notification IDs are typically shorter than group keys.

**Reply by mooseyboots** [reply]
ok, great. thanks for clarifying. (i see i was confusing notif ids with status ids in my data). so we clients can just use what the server returns.

glad this is sorted out!

**Reply by mooseyboots** [reply]
confirming the api endpoints are working in my client now. thanks.
